### ISSUE TITLE
Issue when using clang-17 to compile

### ISSUE THREAD

**Issue by chriscpritchard** [systemd-cron/systemd-cron#141]
I'm using clang version 17 with libcxx and when attempting to compile using clang, I get the following error:

```cp /home/chris/dev/systemd-cron-2.3.0/src/include/libvoreutils.hpp /home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp
clang++ -O2 -Wall -Wextra -fno-exceptions -Wno-psabi  -std=c++20 -I /home/chris/dev/systemd-cron-2.3.0/out/build/include            /home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp -o /home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp.gch
clang++: error: cannot specify -o when generating multiple output files
make: *** [Makefile:220: /home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp.gch] Error 1
```

This does not occur when using gcc and g++ to compile.

**Comment by chriscpritchard** [comment]
Log files from both builds:
```
$ CC="gcc" CXX="g++" ./configure --prefix=/usr --mandir=/usr/share/man --unitdir=/usr/lib/systemd/system --generatordir=/usr/lib/systemd/system-generators --enable-boot=no --enable-minutely=no --enable-runparts=yes --enable-yearly=no --enable-quarterly=no --enable-semi_annually=no
$ CC="gcc" CXX="g++" make
```
[systemd-cron-2.3.0-clang.log](https://github.com/systemd-cron/systemd-cron/files/13418904/systemd-cron-2.3.0-clang.log)
```
$ ./configure --prefix=/usr --mandir=/usr/share/man --unitdir=/usr/lib/systemd/system --generatordir=/usr/lib/systemd/system-generators --enable-boot=no --enable-minutely=no --enable-runparts=yes --enable-yearly=no --enable-quarterly=no --enable-semi_annually=no
$ make
```
[systemd-cron-2.3.0-gcc.log](https://github.com/systemd-cron/systemd-cron/files/13418905/systemd-cron-2.3.0-gcc.log)

**Comment by chriscpritchard** [comment]
That's so bizarre, I've tried removing the space, doesn't seem to work. 

If you have the time can you try running the command with -### as a flag to CLANG? I'm getting the following:

```
clang++ -### -Wall -Wextra -fno-exceptions -Wno-psabi  -std=c++20 -I /home/chris/dev/systemd-cron-2.3.0/out/build/include            /home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp -o /home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp.gch
clang version 17.0.5
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm/17/bin
Configuration file: /etc/clang/x86_64-pc-linux-gnu-clang++.cfg
clang++: error: cannot specify -o when generating multiple output files
 "/usr/lib/llvm/17/bin/clang-17" "-cc1" "-triple" "x86_64-pc-linux-gnu" "-emit-pch" "-dumpdir" "/home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp.gch-" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "libvoreutils.hpp" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/home/chris/dev/systemd-cron-2.3.0" "-resource-dir" "/usr/lib/llvm/17/bin/../../../../lib/clang/17" "-include" "/usr/include/gentoo/fortify.h" "-include" "/usr/include/gentoo/maybe-stddefs.h" "-I" "/home/chris/dev/systemd-cron-2.3.0/out/build/include" "-internal-isystem" "/usr/include/c++/v1" "-internal-isystem" "/usr/lib/llvm/17/bin/../../../../lib/clang/17/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/include" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-Wall" "-Wextra" "-Wno-psabi" "-std=c++20" "-fdeprecated-macro" "-fdebug-compilation-dir=/home/chris/dev/systemd-cron-2.3.0" "-ferror-limit" "19" "-stack-protector" "2" "-fstack-clash-protection" "-fgnuc-version=4.2.1" "-fno-implicit-modules" "-fcolor-diagnostics" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "/home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp.gch" "-x" "c++-header" "/home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp"
 "/usr/lib/llvm/17/bin/ld.lld" "-pie" "--hash-style=gnu" "--eh-frame-hdr" "-m" "elf_x86_64" "-dynamic-linker" "/lib64/ld-linux-x86-64.so.2" "-o" "/home/chris/dev/systemd-cron-2.3.0/out/build/include/libvoreutils.hpp.gch" "/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/Scrt1.o" "/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/crti.o" "/usr/lib/llvm/17/bin/../../../../lib/clang/17/lib/linux/clang_rt.crtbegin-x86_64.o" "-L/usr/lib/gcc/x86_64-pc-linux-gnu/13" "-L/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64" "-L/lib/../lib64" "-L/usr/lib/../lib64" "-L/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/lib" "-L/lib" "-L/usr/lib" "-z" "relro" "-lc++" "-lm" "/usr/lib/llvm/17/bin/../../../../lib/clang/17/lib/linux/libclang_rt.builtins-x86_64.a" "--as-needed" "-lunwind" "--no-as-needed" "-lc" "/usr/lib/llvm/17/bin/../../../../lib/clang/17/lib/linux/libclang_rt.builtins-x86_64.a" "--as-needed" "-lunwind" "--no-as-needed" "/usr/lib/llvm/17/bin/../../../../lib/clang/17/lib/linux/clang_rt.crtend-x86_64.o" "/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/crtn.o"
```
And it might be helpful to try to compare outputs. It's completely understandable if you don't have the time though!

**Comment by nabijaczleweli** [comment]
Using the official clang trunk packages (https://apt.llvm.org/):
```
$ c++ -### -O2 -g -Wall -Wextra -fno-exceptions -Wno-psabi  -std=c++20 -I /home/nabijaczleweli/code/systemd-cron/out/build/include /home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp -o /home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp.gch
Debian clang version 18.0.0 (++20231006112712+a16f6462d756-1~exp1~20231006112827.256)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /bin
 (in-process)
 "/usr/lib/llvm-18/bin/clang" "-cc1" "-triple" "x86_64-pc-linux-gnu" "-emit-pch" "-dumpdir" "/home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp.gch-" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "libvoreutils.hpp" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=none" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debug-info-kind=constructor" "-dwarf-version=5" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/home/nabijaczleweli/code/systemd-cron" "-fcoverage-compilation-dir=/home/nabijaczleweli/code/systemd-cron" "-resource-dir" "/usr/lib/llvm-18/lib/clang/18" "-I" "/home/nabijaczleweli/code/systemd-cron/out/build/include" "-internal-isystem" "/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12" "-internal-isystem" "/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/x86_64-linux-gnu/c++/12" "-internal-isystem" "/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/backward" "-internal-isystem" "/usr/lib/llvm-18/lib/clang/18/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../x86_64-linux-gnu/include" "-internal-externc-isystem" "/usr/include/x86_64-linux-gnu" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-O2" "-Wall" "-Wextra" "-Wno-psabi" "-std=c++20" "-fdeprecated-macro" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fno-implicit-modules" "-fcolor-diagnostics" "-vectorize-loops" "-vectorize-slp" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "/home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp.gch" "-x" "c++-header" "/home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp"
```
it's very odd to me that you're somehow getting a linker cmdline in a compilation step (indeed, if I add a useless `-fuse-ld=lld` clang does complain).

**Comment by nabijaczleweli** [comment]
With /home/nabijaczleweli/code/systemd-cron and /home/chris/dev/systemd-cron-2.3.0 => SYSDUSSY:
```diff
--- chris	2023-11-21 17:55:20.594305868 +0100
+++ nab	2023-11-21 17:55:47.992477806 +0100
@@ -1,10 +1,10 @@
-/usr/lib/llvm/17/bin/clang-17
+/usr/lib/llvm-18/bin/clang
 -cc1
 -triple
 x86_64-pc-linux-gnu
 -emit-pch
 -dumpdir
-SYSDUSSY/out/build/include/libvoreutils.hpp.gch-
+/home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp.gch-
 -disable-free
 -clear-ast-before-backend
 -disable-llvm-verifier
@@ -16,7 +16,7 @@
 -pic-level
 2
 -pic-is-pie
--mframe-pointer=all
+-mframe-pointer=none
 -fmath-errno
 -ffp-contract=on
 -fno-rounding-math
@@ -26,45 +26,50 @@
 x86-64
 -tune-cpu
 generic
+-debug-info-kind=constructor
+-dwarf-version=5
 -debugger-tuning=gdb
--fcoverage-compilation-dir=SYSDUSSY
+-fdebug-compilation-dir=/home/nabijaczleweli/code/systemd-cron
+-fcoverage-compilation-dir=/home/nabijaczleweli/code/systemd-cron
 -resource-dir
-/usr/lib/llvm/17/bin/../../../../lib/clang/17
--include
-/usr/include/gentoo/fortify.h
--include
-/usr/include/gentoo/maybe-stddefs.h
+/usr/lib/llvm-18/lib/clang/18
 -I
-SYSDUSSY/out/build/include
+/home/nabijaczleweli/code/systemd-cron/out/build/include
 -internal-isystem
-/usr/include/c++/v1
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12
 -internal-isystem
-/usr/lib/llvm/17/bin/../../../../lib/clang/17/include
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/x86_64-linux-gnu/c++/12
+-internal-isystem
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/backward
+-internal-isystem
+/usr/lib/llvm-18/lib/clang/18/include
 -internal-isystem
 /usr/local/include
 -internal-isystem
-/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/include
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../x86_64-linux-gnu/include
+-internal-externc-isystem
+/usr/include/x86_64-linux-gnu
 -internal-externc-isystem
 /include
 -internal-externc-isystem
 /usr/include
+-O2
 -Wall
 -Wextra
 -Wno-psabi
 -std=c++20
 -fdeprecated-macro
--fdebug-compilation-dir=SYSDUSSY
 -ferror-limit
 19
--stack-protector
-2
--fstack-clash-protection
 -fgnuc-version=4.2.1
 -fno-implicit-modules
 -fcolor-diagnostics
+-vectorize-loops
+-vectorize-slp
+-faddrsig
 -D__GCC_HAVE_DWARF2_CFI_ASM=1
 -o
-SYSDUSSY/out/build/include/libvoreutils.hpp.gch
+/home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp.gch
 -x
 c++-header
-SYSDUSSY/out/build/include/libvoreutils.hpp
+/home/nabijaczleweli/code/systemd-cron/out/build/include/libvoreutils.hpp
```
(and sorted:
```diff
--- chris.s	2023-11-21 17:57:30.527549006 +0100
+++ nab.s	2023-11-21 17:58:10.093295509 +0100
@@ -1,16 +1,22 @@
 19
 2
-2
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/backward
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/x86_64-linux-gnu/c++/12
+/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../x86_64-linux-gnu/include
 -cc1
 c++-header
 -clear-ast-before-backend
 -debugger-tuning=gdb
+-debug-info-kind=constructor
 -D__GCC_HAVE_DWARF2_CFI_ASM=1
 -disable-free
 -disable-llvm-verifier
 -discard-value-names
 -dumpdir
+-dwarf-version=5
 -emit-pch
+-faddrsig
 -fcolor-diagnostics
 -fcoverage-compilation-dir=SYSDUSSY
 -fdebug-compilation-dir=SYSDUSSY
@@ -21,15 +27,15 @@
 -fmath-errno
 -fno-implicit-modules
 -fno-rounding-math
--fstack-clash-protection
 -funwind-tables=2
 generic
 -I
--include
--include
 /include
 -internal-externc-isystem
 -internal-externc-isystem
+-internal-externc-isystem
+-internal-isystem
+-internal-isystem
 -internal-isystem
 -internal-isystem
 -internal-isystem
@@ -37,14 +43,14 @@
 libvoreutils.hpp
 -main-file-name
 -mconstructor-aliases
--mframe-pointer=all
+-mframe-pointer=none
 -mrelocation-model
 -o
+-O2
 pic
 -pic-is-pie
 -pic-level
 -resource-dir
--stack-protector
 -std=c++20
 SYSDUSSY/out/build/include
 SYSDUSSY/out/build/include/libvoreutils.hpp
@@ -54,14 +60,13 @@
 -triple
 -tune-cpu
 /usr/include
-/usr/include/c++/v1
-/usr/include/gentoo/fortify.h
-/usr/include/gentoo/maybe-stddefs.h
-/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/include
-/usr/lib/llvm/17/bin/clang-17
-/usr/lib/llvm/17/bin/../../../../lib/clang/17
-/usr/lib/llvm/17/bin/../../../../lib/clang/17/include
+/usr/include/x86_64-linux-gnu
+/usr/lib/llvm-18/bin/clang
+/usr/lib/llvm-18/lib/clang/18
+/usr/lib/llvm-18/lib/clang/18/include
 /usr/local/include
+-vectorize-loops
+-vectorize-slp
 -Wall
 -Wextra
 -Wno-psabi
```
)

Which all looks like noise (and my clang defaults to DWARF5, and my clang rembered `-O2`).

**Comment by chriscpritchard** [comment]
I fixed the build by disabling precompiled headers with clang in the makefile as per the below patch. Happy to make a pull request if neccesary, as I don't think the precompiled headers make building this that much quicker? Unless I'm completely missing the point.

```
diff --git a/Makefile.in b/Makefile.in
index 6797c55..19a09c7 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -206,17 +206,10 @@ $(builddir)/include/%.hpp: $(srcdir)/include/%.hpp
        cp $< $@
 
 CXXVER := $(shell $(CXX) --version | { read -r l; echo "$$l"; })
-ifneq "$(findstring clang,$(CXXVER))" ""
-       # clang doesn't use PCHs automatically
-       PCH_ARG := -include-pch $(builddir)/include/libvoreutils.hpp.gch -Wno-gcc-compat
-else
-       PCH_ARG :=
-endif
+PCH_ARG :=
 
-common_headers := $(builddir)/include/configuration.hpp $(builddir)/include/libvoreutils.hpp.gch $(builddir)/include/util.hpp
+common_headers := $(builddir)/include/configuration.hpp $(builddir)/include/libvoreutils.hpp $(builddir)/include/util.hpp
 CFLAGS += -Wall -Wextra -fno-exceptions -Wno-psabi
-$(builddir)/include/libvoreutils.hpp.gch : $(builddir)/include/libvoreutils.hpp
-       $(CXX) $(CFLAGS) $(CPPFLAGS) -std=c++20 -I $(builddir)/include            $< -o $@
 
 $(builddir)/bin/systemd-crontab-generator: $(srcdir)/bin/systemd-crontab-generator.cpp $(common_headers) $(builddir)/include/part2timer.hpp $(builddir)/include/crond2timer.hpp
        $(CXX) $(CFLAGS) $(CPPFLAGS) -std=c++20 -I $(builddir)/include $(PCH_ARG) $< -o $@ $(LDFLAGS) @libcrypto@
```

**Comment by nabijaczleweli** [comment]
Re-measuring this, I see what I expected:
```
$ hyperfine -p 'make clean' 'make -j24' 'make -j25 PCH='
Benchmark 1: make -j24
  Time (mean ± σ):      7.866 s ±  0.086 s    [User: 8.553 s, System: 0.514 s]
  Range (min … max):    7.782 s …  8.056 s    10 runs

Benchmark 2: make -j25 PCH=
  Time (mean ± σ):      7.167 s ±  0.141 s    [User: 9.457 s, System: 0.481 s]
  Range (min … max):    6.939 s …  7.418 s    10 runs

Summary
  'make -j25 PCH=' ran
    1.10 ± 0.02 times faster than 'make -j24'
```
```
$ hyperfine -p 'touch src/bin/crontab.cpp src/bin/systemd-crontab-generator.cpp' 'make -j24' 'make -j25 PCH='
Benchmark 1: make -j24
  Time (mean ± σ):      5.945 s ±  0.698 s    [User: 6.544 s, System: 0.305 s]
  Range (min … max):    5.626 s …  7.912 s    10 runs

Benchmark 2: make -j25 PCH=
  Time (mean ± σ):      7.084 s ±  0.111 s    [User: 9.211 s, System: 0.372 s]
  Range (min … max):    6.956 s …  7.361 s    10 runs

Summary
  'make -j24' ran
    1.19 ± 0.14 times faster than 'make -j25 PCH='
```
```
$ hyperfine -p 'make clean' 'make' 'make PCH='
Benchmark 1: make
  Time (mean ± σ):      9.106 s ±  0.109 s    [User: 8.582 s, System: 0.526 s]
  Range (min … max):    8.976 s …  9.338 s    10 runs

Benchmark 2: make PCH=
  Time (mean ± σ):      9.928 s ±  0.124 s    [User: 9.445 s, System: 0.485 s]
  Range (min … max):    9.783 s … 10.154 s    10 runs

Summary
  'make' ran
    1.09 ± 0.02 times faster than 'make PCH='
```

It's very much worth it for incremental builds (and this is why I originally implemented it), as well as for serial once-over builds, but the PCH blocks the binaries for longer than you get back with parallel once-over builds.

IMO fix your compiler, but failing that, since 1662b899b206f00face30b9d4671551427262b07 you can build with `make PCH=` to disable using the PCH.

**Comment by chriscpritchard** [comment]
Thank you for adding that! I'll mark this as closed.
